common/background_exception_catcher.h

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+// BoB robotics includes
+#include "assert.h"
+
+// Standard C includes
+#include <csignal>
+
+// Standard C++ includes
+#include <array>
+#include <stdexcept>
+#include <string>
+#include <unordered_set>
+
+namespace BoBRobotics {
+//----------------------------------------------------------------------------
+// BoBRobotics::BackgroundExceptionCatcher
+//----------------------------------------------------------------------------
+//! A wrapper for passing exceptions between threads (i.e. a background thread to the main one)
+class BackgroundExceptionCatcher
```

Comment:
You could use the classic Singleton (anti-)design pattern here rather than having so many statics. Also MAYBE the templatedness could even prevent the need for the #define to prevent multiple symbol declaration.
```c++
template<typename T>
class Singleton
{
public:
    static T *getInstance()
    {
        if(s_Instance == nullptr) {
            s_Instance = new T;
        }
        return s_Instance;
    }
private:
    
    static T *s_Instance;
};
Singleton<T>::s_Instance = nullptr;
```

Then you'd use it with the curiously recurring template pattern so:
```c++
class BackgroundExceptionCatcher : public Singleton<BackgroundExceptionCatcher>
{
....
}
```

And access the background exception catcher with ``BackgroundExceptionCatcher::getInstance()``

Reply:
Good idea. I've found a slightly more C++11-y version of this design pattern to use: https://stackoverflow.com/questions/1008019/c-singleton-design-pattern